### PR TITLE
Allow datetime values to be treated as strings

### PR DESCRIPTION
**Pull request by rogordon01**
Currently, default Newtonsoft JsonSerialization settings are used when performing Json deserialization. This impacts datetime values, as the default behavior is to convert them into .Net `DateTime` objects. This causes unwanted formatting issues and leads to loss of TimeZone information on the converted data. 

This PR allows end users to instruct Newtonsoft to simply treat dates as strings during deserialization. This allows the datetime information to be preserved as-is. They can then be assured that any filters are applied on the original value, and not a value modified by Newtownsoft. 

End users must opt-in to this new behavior by specifying a new Parameter.

Current behavior:
image

When enabled:
image

## Description
Describe the changes in this PR.

## Related issues
Addresses [AB#120779](https://microsofthealth.visualstudio.com/Health/_workitems/edit/120779).

## Testing
Describe how this change was tested.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

**Review by ChaChaChaChami — APPROVED** [review]
Makes sense.

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Operations/ConvertData/ConvertDataEngineTests.cs [hunk 1]
```diff
@@ -92,7 +93,7 @@ public async Task GivenCcdaConvertDataRequest_WithADefaultTemplates_CorrectResul
         public async Task GivenJsonConvertDataRequest_WithADefaultTemplates_CorrectResultShouldReturn()
         {
             var convertDataEngine = GetDefaultEngine();
-            var request = GetJsonRequestWithDefaultTemplates();
+            var request = GetJsonRequestWithDefaultTemplates(true);
```
Comment (pallar-ms) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Operations/ConvertData/ConvertDataEngineTests.cs:96]: nit: add param name for values passed as is.

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ConvertDataController.cs [hunk 2]
```diff
@@ -76,6 +78,7 @@ public async Task<IActionResult> ConvertData([FromBody] Parameters inputParams)
             string inputData = ReadStringParameter(inputParams, ConvertDataProperties.InputData);
             string templateCollectionReference = ReadStringParameter(inputParams, ConvertDataProperties.TemplateCollectionReference);
             string rootTemplate = ReadStringParameter(inputParams, ConvertDataProperties.RootTemplate);
+            bool treatDatesAsStrings = ReadBoolParameter(inputParams, ConvertDataProperties.JsonDeserializationTreatDatesAsStrings);
```
Comment (pallar-ms) [inline src/Microsoft.Health.Fhir.Shared.Api/Controllers/ConvertDataController.cs:81]: Just to confirm, this won't error if the request doesn't have this param present?

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ConvertDataController.cs [hunk 3]
```diff
@@ -180,6 +189,7 @@ private static HashSet<string> GetSupportedParams()
                 ConvertDataProperties.InputDataType,
                 ConvertDataProperties.TemplateCollectionReference,
                 ConvertDataProperties.RootTemplate,
+                ConvertDataProperties.JsonDeserializationTreatDatesAsStrings,
```
Comment (pallar-ms) [inline src/Microsoft.Health.Fhir.Shared.Api/Controllers/ConvertDataController.cs:192]: By adding it in the supported param list, upon validation of params we are enforcing this parameter is provided in the request, right? Is that intended or can this simply not be provided in the request and default if so. If it is the former, I suppose we will need an api version update.

Comment (rogordon01) [reply src/Microsoft.Health.Fhir.Shared.Api/Controllers/ConvertDataController.cs:192]: The supported param list contains all of the params that this controller works with. So if a parameter provided by the end user doesn't exist within this list then the user will get an error. It doesn't enforce that these params need to _exist_, though. The various `ReadXXXParameter` method seem to do that.  

As you've stated, I've opted to allow a default value to be used if one isn't provided. I added a new `ReadBoolParameter` which will return `false` if the parameter is not present.

**Review by pallar-ms — DISMISSED** [review]

File: src/Microsoft.Health.Fhir.Core/Data/OperationDefinition/convert-data.json [hunk 4]
```diff
@@ -43,6 +43,14 @@
             "max": "1",
             "documentation": "The root template to use while transforming the data. Supported default values - `ADT_A01`, `OML_O21`, `ORU_R01`, `VXU_V04` for input data type `Hl7v2`, `CCD`, `ConsultationNote`, `DischargeSummary`, `HistoryandPhysical`, `OperativeNote`, `ProcedureNote`, `ProgressNote`, `ReferralNote`, `TransferSummary` for input data type `Ccda`, `ExamplePatient`, `Stu3ChargeItem` for data type `Json`, and use resource name as root template for `Fhir`",
             "type": "string"
+        },
+        {
+            "name": "jsonDeserializationTreatDatesAsStrings",
+            "use": "in",
+            "min": 1,
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Data/OperationDefinition/convert-data.json:50]: Since this isn't a required field the min should be 0.

Comment (rogordon01) [reply src/Microsoft.Health.Fhir.Core/Data/OperationDefinition/convert-data.json]: Done

**Review by LTA-Thinking — APPROVED** [review]

**Pull request merged**